src/ray/gcs/gcs_server/gcs_node_manager.h

```diff
@@ -171,8 +176,30 @@ class GcsNodeManager : public rpc::NodeInfoHandler {
   /// \param node The node which is dead.
   void AddDeadNodeToCache(std::shared_ptr<rpc::GcsNodeInfo> node);
 
+  /// Set the death info of the node.
+  ///
+  /// \param node_id The ID of the node.
+  /// \param death_info The death info of the node, must be non-null.
+  void SetDeathInfo(const NodeID &node_id, shared_ptr<rpc::NodeDeathInfo> death_info);
+
+  /// Infer death cause of the node based on existing draining requests.
+  ///
+  /// \param[in] node_id The ID of the node. The node must not be removed
+  /// from alive nodes yet.
+  /// \param[out] caused_by_draining The inferred result that whether the node death
+  /// is caused by draining.
+  /// \return The inferred death info of the node, non-null.
+  shared_ptr<rpc::NodeDeathInfo> InferDeathInfo(const NodeID &node_id,
+                                                bool &caused_by_draining);
```

Comment:
hmm, it's wried that this function has two return values: one is normal return and one is out param.

Seems you don't need to return caused_by_draining. You can just check NodeDeathInfo.reason  == rpc::NodeDeathInfo::UNEXPECTED_TERMINATION to determine `is_intended`?

Reply:
Good call.